python/tests/test_topology.py

```diff
@@ -917,89 +917,96 @@ def ts_kc_distance_incremental(ts1, ts2, lambda_=0):
-    # by inserted edges.
-    for e in edges_out:
-        update_in_subtree_pairs(tree, kc, e.child, sample_index_map)
-
-    for e in edges_in:
-        if tree.is_leaf(e.child):
-            u = e.child
+    # Update state of detached subtrees.
+    for e in reversed(edges_out):
+        u = e.child
+        depths[u] = 0
+
+        # Only update detached subtrees that remain detached. Otherwise,
+        # they must be reattached by an incoming edge and will be
+        # updated below.
+        if tree.parent(u) == tskit.NULL:
+            update_kc_subtree_state(tree, kc, u, sample_index_map, depths)
+
+    # Propagate state change down into reattached subtrees.
+    for e in reversed(edges_in):
+        u = e.child
+        assert depths[u] == 0
+        depths[u] = depths[e.parent] + 1
+        update_kc_subtree_state(tree, kc, u, sample_index_map, depths)
+
+        # The per-leaf elements of KC only change when the edge directly
+        # above the leaf changes, so are handled separately from the
+        # propagated state used for leaf-pair elements.
+        if tree.is_leaf(u):
             time = tree.branch_length(u)
             update_kc_vectors_single_leaf(kc, u, time, sample_index_map)
-        update_in_out_subtree_pairs(tree, kc, e, sample_index_map)
-        update_in_subtree_pairs(tree, kc, e.child, sample_index_map)
 
 
-def update_in_subtree_pairs(tree, kc, u, sample_index_map):
+def update_kc_subtree_state(tree, kc, u, sample_index_map, depths):
     """
-    Visit all pairs of leaves of the internal node u, tracking
-    the mrca of each pair and the depth of that mrca.
+    Update the depths of the nodes in this subtree. When a leaf is hit,
+    update the KC vector elements associated with that leaf.
     """
-    root_time = tree.time(tree.root)
-    stack = [(u, tree.depth(u))]
+    stack = [u]
     while len(stack) > 0:
-        v, depth = stack.pop()
-        if not tree.is_leaf(v):
-            c1 = tree.left_child(v)
-            while c1 != -1:
-                stack.append((c1, depth + 1))
-                c2 = tree.right_sib(c1)
-                while c2 != -1:
-                    for l1 in tree.leaves(c1):
-                        for l2 in tree.leaves(c2):
-                            time = root_time - tree.time(v)
-                            update_kc_vectors_pair(
-                                kc, l1, l2, depth, time, sample_index_map
-                            )
-                    c2 = tree.right_sib(c2)
-                c1 = tree.right_sib(c1)
-
-
-def update_in_out_subtree_pairs(tree, kc, edge, sample_index_map):
+        v = stack.pop()
+        if tree.is_leaf(v):
+            update_kc_pairs_with_leaf(tree, kc, v, sample_index_map, depths)
+        else:
+            c = tree.left_child(v)
+            while c != -1:
+                # Terminate iteration at nodes that are currently considered
+                # roots by the edge diffs
```

Comment:
This is a little not obvious but is because we're iterating over the completed tree as opposed to the edge diffs as their being inserted so this is a way to see where we are in the insertion, assuming this is done top-down. Any better way of doing that?